File: rust/src/audio/audio_graph_test.rs
use crate::audio::audio_graph::{AudioGraph, AudioNode};

struct DummyNode {
    name: String,
}

impl DummyNode {
    fn new(name: &str) -> Self {
        Self {
            name: name.to_string(),
        }
    }
}

impl AudioNode for DummyNode {
    fn name(&self) -> &str {
        &self.name
    }

    fn process(&mut self, _buffer: &mut [f32]) {
        // Dummy processing
    }
}

#[test]
fn test_audio_graph_lifecycle() {
    let mut graph = AudioGraph::new();
    graph.start();
    graph.stop();
}

#[test]
fn test_audio_graph_connect() {
    let mut graph = AudioGraph::new();
    graph.add_node(Box::new(DummyNode::new("src")));
    graph.add_node(Box::new(DummyNode::new("dest")));

    assert!(graph.connect("src", "dest").is_ok());
    assert!(graph.connect("unknown", "dest").is_err());
}

File: csharp/src/Audio/AudioGraphTest.cs
using System;

namespace BQt.Audio.Tests
{
    public class AudioGraphTest
    {
        private class DummyNode : IAudioNode
        {
            public string Name { get; }

            public DummyNode(string name)
            {
                Name = name;
            }

            public void Process(float[] buffer)
            {
            }
        }

        public static void Main()
        {
            TestAudioGraphLifecycle();
            TestAudioGraphConnect();
            Console.WriteLine("All tests passed.");
            OmniGainTest.Run();
            OmniSynthesizerTest.Run();
        }

        private static void TestAudioGraphLifecycle()
        {
            var graph = new AudioGraph();
            graph.Start();
            graph.Stop();
        }

        private static void TestAudioGraphConnect()
        {
            var graph = new AudioGraph();
            graph.AddNode(new DummyNode("src"));
            graph.AddNode(new DummyNode("dest"));

            try
            {
                graph.Connect("src", "dest");
            }
            catch (Exception)
            {
                throw new Exception("Test failed: Should be able to connect existing nodes");
            }

            bool threw = false;
            try
            {
                graph.Connect("unknown", "dest");
            }
            catch (ArgumentException)
            {
                threw = true;
            }

            if (!threw)
            {
                throw new Exception("Test failed: Should throw when connecting unknown nodes");
            }
        }
    }
}
